libbpf-tools/ttysnoop.bpf.c

```diff
@@ -0,0 +1,120 @@
+// SPDX-License-Identifier: (LGPL-2.1 OR BSD-2-Clause)
+// Copyright (c) 2023 Kylin
+// Author: Jackie Liu <liuyun01@kylinos.cn>
+//
+// Base on ttysnoop.py 2016 Brendan Gregg.
+//
+#include <vmlinux.h>
+#include <bpf/bpf_core_read.h>
+#include <bpf/bpf_tracing.h>
+#include <bpf/bpf_helpers.h>
+#include "ttysnoop.h"
+#include "compat.bpf.h"
+#include "core_fixes.bpf.h"
+
+#define WRITE		1
+#define ITER_UBUF	6
+
+extern __u32 LINUX_KERNEL_VERSION __kconfig;
+
+const volatile int user_data_count = 16;
+const volatile int pts_inode = -1;
+
+static __always_inline int
+do_tty_write(void *ctx, const char *buf, size_t count)
+{
+	for (int i = 0; i < user_data_count && count; i++) {
+		struct event *event = reserve_buf(sizeof(*event));
+		if (!event)
+			break;
+
+		 /**
+		  * bpf_probe_read_user() can only use a fixed size, so truncate to count
+		  * in user space
+		  */
+		if (bpf_probe_read_user(&event->buf, sizeof(event->buf), (void *)buf)) {
+			discard_buf(event);
+			break;
+		}
+
+		event->count = count > BUFSIZE ? BUFSIZE : count;
+		submit_buf(ctx, event, sizeof(*event));
+
+		if (count < BUFSIZE)
+			break;
+
+		count -= BUFSIZE;
+		buf += BUFSIZE;
+	}
+
+	return 0;
+}
+
+SEC("kprobe/tty_write")
+int BPF_KPROBE(kprobe__tty_write)
+{
+	size_t count = 0;
+	const char *buf = NULL;
+	const struct file *file;
+
+	/*
+	 * commit 9bb48c82aced (v5.11-rc4) tty: implement write_iter
+	 * hanged arguments of tty_write function
+	 */
+	if (LINUX_KERNEL_VERSION < KERNEL_VERSION(5, 10, 11)) {
```

Comment:
Could we avoid this version check using CO-RE ?

Reply:
tty_write changed param after commit 9bb48c82aced, it not a new function name or new structs, I can only think of using KCONFIG to distinguish between different kernels, does we have any other solutions?

Reply:
I suddenly had an idea, maybe we can refer to ksnoop to get the parameter type of the function from the BTF, so that we can observe the currently running kernel version.

Reply:
Sounds good.